File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(n)
    raise RuntimeError if n < 1
    sum = factors(n).reduce(1, :+)
    return 'deficient' if sum < n
    return 'perfect'   if sum == n
    return 'abundant'  if sum > n
  end

  def self.factors(n)
    out = []
    (1..n-1).each do |x|
      (x..n-1).each do |y|
        candidate = x*y
        out.push(x) if candidate == n
        out.push(y) if candidate == n
      end
    end
    out
  end
end

module BookKeeping
  VERSION = 1
end

File: ecmascript/perfect-numbers/perfect-numbers.js
class PerfectNumbers {
  classify(num) {
    if (num < 1) {
      throw 'Classification is only possible for natural numbers.';
    }
    const factors = this.getFactors(num).filter((i) => i != num);
    const sum = factors.reduce((sum, i) => sum + i, 0);
    if (num == sum) { return 'perfect'; }
    if (num <  sum) { return 'abundant'; }
    if (num >  sum) { return 'deficient'; }
    return 'foo';
  }

  // https://stackoverflow.com/a/43204663
  getFactors(num) {
    const isEven = num % 2 === 0;
    let inc = isEven ? 1 : 2;
    let factors = [1, num];

    for (let curFactor = isEven ? 2 : 3; Math.pow(curFactor, 2) <= num; curFactor += inc) {
      if (num % curFactor !== 0) continue;
      factors.push(curFactor);
      let compliment = num / curFactor;
      if (compliment !== curFactor) factors.push(compliment);
    }

    return factors;
  }
}

export default PerfectNumbers;